source/module_basis/module_nao/hydrogen_radials.cpp

```diff
@@ -0,0 +1,412 @@
+    std::map<std::pair<int, int>, std::pair<std::vector<double>, std::vector<double>>> radials; 
+    std::map<std::pair<int, int>, double> rmaxs;
+
+    std::vector<std::pair<int, int>> nl_pairs = unzip_strategy(nmax, strategy);
+    double rmax = 0.0;
+    for(auto nl_pair : nl_pairs)
+    {
+        int n = nl_pair.first;
+        int l = nl_pair.second;
+        std::vector<double> rgrid;
+        std::vector<double> rvalue;
+        double rmax_nl = generate_hydrogen_radial_toconv(charge,
+                                                         n,
+                                                         l,
+                                                         conv_thr,
+                                                         rank,
+                                                         rgrid,
+                                                         rvalue,
+                                                         ptr_log);
+        radials[nl_pair] = std::make_pair(rgrid, rvalue);
+        rmaxs[nl_pair] = rmax_nl;
+        if(rmax < rmax_nl)
+        {
+            rmax = rmax_nl;
+        }
+    }
+    // zero padding towards rmax
+    for(auto& radial : radials)
+    {
+        int n = radial.first.first;
+        int l = radial.first.second;
+        std::pair<int, int> nl_pair = std::make_pair(n, l);
+
+        std::vector<double>& rgrid = radial.second.first;
+        std::vector<double>& rvalue = radial.second.second;
+
+        if(rmaxs[nl_pair] < rmax)
+        {
+            int ngrid = static_cast<int>((rmax - rmaxs[nl_pair]) / dr); // (r, rmax]
+            for(int ir = 1; ir <= ngrid; ++ir)
+            {
+                rgrid.push_back(rmaxs[nl_pair] + ir * dr);
+                rvalue.push_back(0.0);
+            }
+        }
+        // smooth the tail
+        smooth(rgrid, rvalue, 0.1);
+    }
+    return radials;
+}
+
+std::map<std::pair<int, int>, std::pair<int, int>>
+HydrogenRadials::mapping_nl_lzeta(const int nmax,
+                                  const std::string strategy)
+{
+    std::map<std::pair<int, int>, std::pair<int, int>> nl_lzeta;
+    std::vector<std::pair<int, int>> nl_pairs = unzip_strategy(nmax, strategy);
+    // initialize nzetas by all zeros
+    // get lmax
+    int lmax = 0;
+    for(auto nl_pair: nl_pairs)
+    {
+        lmax = lmax>nl_pair.second?lmax:nl_pair.second;
+    }
+    std::vector<int> nzetas(lmax + 1, 0);
+    for(auto nl_pair : nl_pairs)
+    {
+        int n = nl_pair.first;
+        int l = nl_pair.second;
+        nl_lzeta[nl_pair] = std::make_pair(l, nzetas[l]);
+        nzetas[l] += 1;
+    }
+    // import to standard RadialSet data structure
+    lmax_ = lmax;
+    nzeta_ = new int[lmax_ + 1];
+    for(int l = 0; l <= lmax_; ++l)
+    {
+        nzeta_[l] = nzetas[l];
+        nzeta_max_ = nzeta_max_>nzeta_[l]?nzeta_max_:nzeta_[l];
+    }
+    indexing();
+
+    return nl_lzeta;
+}
+
+void HydrogenRadials::hydrogen(const double charge,
```

Comment:
generate orbitals, then map to (l, zeta), then it will be feasible to save to NumericalRadial data carrier.